prov/util/src/util_coll.c

```diff
@@ -0,0 +1,1042 @@
+
+		if (comp_data->cid_buf[iter]) {
+			tmp = comp_data->cid_buf[iter];
+			pos = 0;
+			while (!(tmp & 0x1)) {
+				tmp >>= 1;
+				pos++;
+			}
+
+			/* clear the bit from global cid space */
+			util_coll_cid[iter] ^= (1 << pos);
+			lsb_set_pos += pos;
+		} else {
+			lsb_set_pos += sizeof(comp_data->cid_buf[0]) * 8;
+		}
+	}
+	assert(lsb_set_pos < OFI_CONTEXT_ID_SIZE * 8);
+	coll_mc->cid = lsb_set_pos;
+	coll_mc->tag_seq = 0;
+	free(comp_data);
+
+	/* write to the eq  */
+	memset(&entry, 0, sizeof(entry));
+	entry.fid = &coll_mc->mc_fid.fid;
+	entry.context = coll_mc->mc_fid.fid.context;
+	bytes = sizeof(struct fi_eq_entry);
+
+	ep  = container_of(coll_mc->ep, struct util_ep, ep_fid);
+	if (ofi_eq_write(&ep->eq->eq_fid, FI_JOIN_COMPLETE,
+			 &entry, (size_t) bytes, FI_COLLECTIVE) < 0)
+		FI_WARN(ep->domain->fabric->prov, FI_LOG_DOMAIN,
+			"join collective - eq write failed\n");
+}
+
+void util_coll_barrier_comp(struct util_coll_mc *coll_mc,
+			    struct util_coll_comp_item *comp)
+{
+/* TODO: We should write a completion to the CQ, not the EQ
+	struct fi_eq_err_entry entry;
+	struct util_ep *ep;
+	ssize_t bytes;
+
+	memset(&entry, 0, sizeof(entry));
+	entry.fid = &coll_mc->mc_fid.fid;
+	entry.context = coll_mc->mc_fid.fid.context;
+	bytes = sizeof(struct fi_eq_entry);
+	free(comp->data);
+
+	ep  = container_of(coll_mc->ep, struct util_ep, ep_fid);
+	if (ofi_eq_write(&ep->eq->eq_fid, FI_COLL,
+			 &entry, (size_t) bytes, FI_COLLECTIVE) < 0)
+		FI_WARN(ep->domain->fabric->prov, FI_LOG_DOMAIN,
+			"barrier collective - eq write failed\n");
+*/
+}
+
+/* TODO: We can probably use the atomic defines for these, rather than
+ * duplicate the definitions here
+ */
+#define UTIL_COLL_REDUCE_FUNC(src,dst,cnt,OP,TYPE)	\
+		do { \
+			int i; \
+			TYPE *d = (dst); \
+			TYPE *s = (src); \
+			for (i=0; i<(cnt); i++) {\
+				OP(d[i],s[i]); \
+			} \
+		} while (0)
+
+#define SWITCH_REDUCE(type,FUNC,...)		\
+switch (type) {							 \
+		case FI_INT8:	FUNC(__VA_ARGS__,int8_t); break; \
+		case FI_UINT8:	FUNC(__VA_ARGS__,uint8_t); break; \
+		case FI_INT16:	FUNC(__VA_ARGS__,int16_t); break; \
+		case FI_UINT16: FUNC(__VA_ARGS__,uint16_t); break; \
+		case FI_INT32:	FUNC(__VA_ARGS__,int32_t); break; \
+		case FI_UINT32: FUNC(__VA_ARGS__,uint32_t); break; \
+		case FI_INT64:	FUNC(__VA_ARGS__,int64_t); break; \
+		case FI_UINT64: FUNC(__VA_ARGS__,uint64_t); break; \
+		default: return -FI_EOPNOTSUPP; \
+		}
+
+#define UTIL_COLL_MIN(dst,src)	if ((dst) > (src)) (dst) = (src)
+#define UTIL_COLL_MAX(dst,src)	if ((dst) < (src)) (dst) = (src)
+#define UTIL_COLL_SUM(dst,src)	(dst) += (src)
+#define UTIL_COLL_PROD(dst,src)	(dst) *= (src)
+#define UTIL_COLL_LOR(dst,src)	(dst) = (dst) || (src)
+#define UTIL_COLL_LAND(dst,src)	(dst) = (dst) && (src)
+#define UTIL_COLL_BOR(dst,src)	(dst) |= (src)
+#define UTIL_COLL_BAND(dst,src)	(dst) &= (src)
+#define UTIL_COLL_LXOR(dst,src)	(dst) = ((dst) && !(src)) || (!(dst) && (src))
+#define UTIL_COLL_BXOR(dst,src)	(dst) ^= (src)
```

Comment:
util_atomic.c defines these macros.  See OFI_OP_XXX(type,dst,src).  Those should be used instead.